packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp

```diff
@@ -631,6 +631,7 @@ namespace Ifpack2 {
 #ifdef HAVE_IFPACK2_MPI
         for (local_ordinal_type i=0,iend=pids.recv.extent(0);i<iend;++i)
           MPI_Cancel(&reqs.recv[i]);
+        waitall(reqs.send.size(), reqs.send.data());
```

Comment:
That's right, you have to wait on cancelled messages.

Reply:
Is it? Here I'm waiting on the sends, not the receives.

I believe MPI_Cancel, MPI_Waitany, and MPI_Waitall have the same semantics w.r.t. termination of a request. But I definitely could be wrong. Did some googling and still think a wait isn't needed on a cancelled request.

The motivation for waiting on the sends is that (1) that should have been done and I missed it; (2) I think the sends are not done when the next test, using the same BlockTriDi object (a good thing) starts, after the norm-based termination test. The motivation not to cancel the sends is that MPI_Cancel warns against it.

For clarity, all of this concerns the overlapped-comm code path, which in practice isn't being used.

Reply:
I just confirmed that there is a comm issue in the overlapped-comm code path where norm based termination is on. I agree that it is not important and it is not used in practice. I am fine if we decide not to fix now.

Reply:
I'd like to fix it. I _think_ the above commit should. Since you have a reproducer, can you check if inserting the waitall line above fixes it? I couldn't reproduce it, so I don't know for sure that it fixes the problem.

Reply:
@ambrad MPI standard says: "It is still necessary to call `MPI_REQUEST_FREE`, `MPI_WAIT` or `MPI_TEST` (or any of the derived operations) with the cancelled request as argument after the call to `MPI_CANCEL`."

Reply:
@ambrad `MPI_WAIT` "behaves as a local function" (i.e., "is guaranteed to return, irrespective of the activities of other processes") if called on a cancelled thing.  Also, "if `MPI_TEST` is repeatedly called in a busy wait loop for a cancelled communication, then `MPI_TEST` will eventually be successful."

Reply:
A few things.

First, I'm speculating the OpenMP builds may be failing b/c there's processor imbalance, which makes MPI incorrectness exibit itself with greater likelihood.

Second, after finding a better source to read about MPI_Cancel, I think I want to remove it completely, and simply do
```
        waitall(reqs.recv.size(), reqs.recv.data());
        waitall(reqs.send.size(), reqs.send.data());
```
This will let the final round of comm complete, but nothing will be done with the data. The more I read about MPI_Cancel, the more I think it's not going to provide any speedup to cancel the receives. In any case, the speedup is minor, a fraction of one iteration at most.

Third, after reviewing the unit tests, I decided I want to test the advanced features in the general test (i.e., `{"general", false, false}, // Test the general case.`) in every parameterization of BlockTriDi. Currently, these do not get tested if BlockRelaxation wraps BlockTriDi, and this happens whenever the feature set to be tested falls within BlockRelaxation's feature set. So I'm adjusting the unit tester to _always_ test the advanced features. The unit tests are fast, so while this adds ~50% (very rough estimate), this is to a ~1-2s overall test time, which I view as insignificant.

Reply:
@ambrad That makes sense -- the MPI Standard is full of warnings about how `MPI_Cancel` isn't fast, etc.